src/react/client.tsx

```diff
@@ -113,6 +115,27 @@ export function AuthProvider({
     [storageSet, storageRemove],
   );
 
+  useEffect(() => {
+    const listener = async (e: Event) => {
+      if (isRefreshingToken) {
+        // There are 3 different ways to trigger this pop up so just try all of
+        // them.
+
+        e.preventDefault();
+        // This confirmation message doesn't actually appear in most modern
+        // browsers but we tried.
+        const confirmationMessage =
+          "Are you sure you want to leave? Your changes may not be saved.";
+        e.returnValue = true;
+        return confirmationMessage;
+      }
+    };
+    browserAddEventListener("beforeunload", listener);
+    return () => {
+      browserRemoveEventListener("beforeunload", listener);
+    };
+  });
```

Comment:
no deps means this runs every render, which works but could be expensive if this provider isn't at the top